source/simulator/introspection.cc

```diff
@@ -166,6 +166,90 @@ namespace aspect
   }
 
 
+  /**
+   * This is an internal deal.II function stolen from dof_tools.cc
+   */
+  template <int dim, int spacedim>
+  std::vector<unsigned char>
+  get_local_component_association (const FiniteElement<dim,spacedim>  &fe,
+                                   const ComponentMask        &component_mask)
+  {
+    std::vector<unsigned char> local_component_association (fe.dofs_per_cell,
+                                                            (unsigned char)(-1));
+
+    // compute the component each local dof belongs to.
+    // if the shape function is primitive, then this
+    // is simple and we can just associate it with
+    // what system_to_component_index gives us
+    for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
+      if (fe.is_primitive(i))
+        local_component_association[i] =
+          fe.system_to_component_index(i).first;
+      else
+        // if the shape function is not primitive, then either use the
+        // component of the first nonzero component corresponding
+        // to this shape function (if the component is not specified
+        // in the component_mask), or the first component of this block
+        // that is listed in the component_mask (if the block this
+        // component corresponds to is indeed specified in the component
+        // mask)
+        {
+          const unsigned int first_comp =
+            fe.get_nonzero_components(i).first_selected_component();
+
+          if ((fe.get_nonzero_components(i)
+               &
+               component_mask).n_selected_components(fe.n_components()) == 0)
+            local_component_association[i] = first_comp;
+          else
+            // pick the component selected. we know from the previous 'if'
+            // that within the components that are nonzero for this
+            // shape function there must be at least one for which the
+            // mask is true, so we will for sure run into the break()
+            // at one point
+            for (unsigned int c=first_comp; c<fe.n_components(); ++c)
+              if (component_mask[c] == true)
+                {
+                  local_component_association[i] = c;
+                  break;
+                }
+        }
+
+    Assert (std::find (local_component_association.begin(),
+                       local_component_association.end(),
+                       (unsigned char)(-1))
+            ==
+            local_component_association.end(),
+            ExcInternalError());
+
+    return local_component_association;
+  }
+
+  template <int dim>
+  IndexSet extract_component_subset(const DoFHandler<dim> &dof_handler,
+                                    const ComponentMask &component_mask)
```

Comment:
I don't like the name. Can we call it `extract_dofs_for_component_subset()` or something?